cmd/federatedrole/apply.go

```diff
@@ -0,0 +1,199 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *applyOptions) run() error {
+	ctx := context.TODO()
+
+	var (
+		input     io.Reader
+		err       error
+		accountID string
+		uid       string
+		ok        bool
+	)
+	if o.url != "" {
+		resp, err := http.Get(o.url)
+		if err != nil {
+			return err
+		}
+
+		if resp.StatusCode/100 != 2 {
+			return errors.New("Failed to send GET request to " + o.url)
+		}
+
+		defer resp.Body.Close()
+		input = resp.Body
+	} else {
+		path, err := filepath.Abs(o.file)
+		if err != nil {
+			return err
+		}
+
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		input = file
+	}
+
+	var federatedRole awsv1alpha1.AWSFederatedRole
+	d := yaml.NewYAMLOrJSONDecoder(input, 4096)
+	if err := d.Decode(&federatedRole); err != nil {
+		return err
+	}
+
+	awsSetupClient, err := awsprovider.NewAwsClient(o.profile, o.region, o.cfgFile)
+	if err != nil {
+		return err
+	}
+
+	var federatedAccountAccesses awsv1alpha1.AWSFederatedAccountAccessList
+	if err := o.kubeCli.List(ctx, &federatedAccountAccesses, &client.ListOptions{}); err != nil {
+		return err
+	}
+
+	callerIdentityOutput, err := awsSetupClient.GetCallerIdentity(&sts.GetCallerIdentityInput{})
+	if err != nil {
+		klog.Error("Fail to get caller identity. Could you please validate the credentials?")
+		return err
+	}
+	if o.verbose {
+		fmt.Fprintln(o.Out, callerIdentityOutput)
+	}
+
+	if len(federatedAccountAccesses.Items) == 0 {
+		fmt.Fprintln(o.Out, "Cannot find associated AWS federated account accesses")
+		return nil
+	}
+
+	for _, federatedAccount := range federatedAccountAccesses.Items {
+		if federatedAccount.Spec.AWSFederatedRole.Namespace == federatedRole.Namespace &&
+			federatedAccount.Spec.AWSFederatedRole.Name == federatedRole.Name {
+			if accountID, ok = federatedAccount.Labels[awsAccountIDLabel]; !ok {
+				return errors.New("Unable to get AWS AccountID label for AWS federated account access CR " + federatedAccount.Name)
+			}
+
+			if uid, ok = federatedAccount.Labels[uidLabel]; !ok {
+				return errors.New("Unable to get UID label for AWS federated account access CR " + federatedAccount.Name)
+			}
+
+			roleArn := fmt.Sprintf("arn:aws:iam::%s:role/%s", accountID, awsv1alpha1.AccountOperatorIAMRole)
+			credentials, err := awsprovider.GetAssumeRoleCredentials(awsSetupClient, aws.Int64(900),
+				callerIdentityOutput.UserId, &roleArn)
+			if err != nil {
+				return err
+			}
+
+			awsClient, err := awsprovider.NewAwsClientWithInput(&awsprovider.AwsClientInput{
+				AwsIDKey:     *credentials.AccessKeyId,
+				AwsAccessKey: *credentials.SecretAccessKey,
+				AwsToken:     *credentials.SessionToken,
+				AwsRegion:    o.region,
+			})
+			if err != nil {
+				return err
+			}
+
+			fmt.Fprintln(o.Out, "Updating IAM policy for AWS federated account access CR "+federatedAccount.Name)
+			if err := awsprovider.RefreshIAMPolicy(awsClient, &federatedRole, accountID, uid); err != nil {
```

Comment:
One case I just thought of is the FederatedRole CR may not have a new custom policy but have a new managed Policy in the list. So the code should also ensure that all the Managed policies specified are attached to the Role and detached any that may have been removed